src/include/index/flatpq_index.h

```diff
@@ -0,0 +1,927 @@
+        //          << std::get<0>(high_scores[i]) << ")" << std::endl;
+
+        auto [degree, zero_part] = low_degrees[i];
+        auto [score, high_vector_id] = high_scores[i];
+        auto low_centroid = new_centroids[zero_part];
+        auto high_vector = training_set[high_vector_id];
+
+        for (size_t i = sub_begin; i < sub_end; ++i) {
+          low_centroid[i] = high_vector[i];
+        }
+        for (size_t i = sub_begin; i < sub_end; ++i) {
+          new_centroids[parts[high_vector_id]][i] -= high_vector[i];
+        }
+
+        ++degrees[zero_part];
+        --degrees[parts[high_vector_id]];
+      }
+    }
+#endif
+
+    // Check for convergence
+    max_diff = 0;
+    total_weight = 0;
+    for (size_t j = 0; j < num_vectors(centroids); ++j) {
+      if (degrees[j] != 0) {
+        auto centroid = new_centroids[j];
+        for (size_t k = sub_begin; k < sub_end; ++k) {
+          centroid[k] /= degrees[j];
+          total_weight += centroid[k] * centroid[k];
+        }
+      }
+      auto diff = sub_sum_of_squares(
+          centroids[j], new_centroids[j], sub_begin, sub_end);
+      max_diff = std::max<double>(max_diff, diff);
+    }
+    centroids.swap(new_centroids);
+
+    if (max_diff < tol * total_weight) {
+      break;
+    }
+  }
+  return std::make_tuple(iter, max_diff / total_weight);
+}
+
+/**
+ * @brief Flat index that uses product quantization
+ *
+ * @tparam T Data type of the input vectors
+ * @tparam shuffled_ids_type Data type of the shuffled ids
+ * @tparam indices_type Data type of the indices
+ */
+template <
+    class T,
+    class shuffled_ids_type = size_t,
+    class indices_type = size_t>
+class flatpq_index {
+  // using feature_type = typename
+  // std::remove_reference_t<decltype(partitioned_db)>::value_type;
+  using feature_type = T;
+  using id_type = shuffled_ids_type;
+  using score_type = float;
+  using centroid_feature_type = float;
+  using code_type = uint8_t;
+
+  // @todo Temporary only!!
+ public:
+  // metadata
+  size_t dimension_{0};
+  size_t num_subspaces_{0};
+  size_t sub_dimension_{0};
+  size_t bits_per_subspace_{8};
+  size_t num_clusters_{256};
+  float tol_ = 0.001;
+  size_t max_iter_ = 16;
+  size_t num_threads_ = std::thread::hardware_concurrency();
+
+  using metadata_element = std::tuple<std::string, void*, tiledb_datatype_t>;
+  std::vector<metadata_element> metadata{
+      {"dimension", &dimension_, TILEDB_UINT64},
+      {"num_subspaces", &num_subspaces_, TILEDB_UINT64},
+      {"sub_dimension", &sub_dimension_, TILEDB_UINT64},
+      {"bits_per_subspace", &bits_per_subspace_, TILEDB_UINT64},
+      {"num_clusters", &num_clusters_, TILEDB_UINT64},
+      {"tol", &tol_, TILEDB_FLOAT32},
+      {"max_iter", &max_iter_, TILEDB_UINT64},
+      {"num_threads", &num_threads_, TILEDB_UINT64},
+  };
+
+  // array data
+  ColMajorMatrix<centroid_feature_type> centroids_;
+  std::vector<ColMajorMatrix<score_type>> distance_tables_;
```

Comment:
What is the expected RAM size for this?

Reply:
Not very much: (score size) * (# centroids)^2 * (# subspaces).  If we take the usual float scores, 8-bit codes, and 8 subspaces, that is 4 * 256 * 256 * 8 = 2MB -- for all of the tables.  Each table is 256KB.